### PR TITLE
Introduce a Page component that can be used to wrap all page co…

### LINKED ISSUE #8703: Revisit the way we address a glitch when navigating from the home page
This is a follow-up to #8662 

Some code landed in (insert link to commit here) which resets the `viewContext` when navigating away from the home page to address an issue where the layout was broken during the transition because the app thought it was still on the home page. 

This approach is problematic. A partial solution to this was suggested at https://github.com/mozilla/addons-frontend/pull/8662#discussion_r330258915, which does address the full-width loading glitch, but did not address the missing margin glitch. It is thought that the latter can be addressed by controlling the margins around the `AppBanner` externally, rather than having the `AppBanner` control its own margins.

Note that I will update this issue with links once the code in question lands.

### PR DESCRIPTION
**Pull request by bobsilverberg**
Fixes #8703 

Here is a better fix for #8703. It undoes the change to the `viewContext` reducer, and it seems to fix all of the problems that were present before. Unfortunately simply wrapping the entire contents of the `render` function inside a `<Page>` component has produced a tremendous amount of churn in most of the `page` files, but for the most part nothing has changed in them other than adding that `<Page>` component.

I did remove a couple of things from `App`, as I couldn't detect what the need for them was. If I need to re-introduce them I can:

1. `App` was creating a property called `this.header`, which was being set to a `ref` for the `Header` component. But I didn't see that property used anywhere. 
2. `App` also had a prop called `HeaderComponent` which seemed to allow a caller to pass a component other than `Header` into `App`. I don't see that used anywhere, not even in any tests, so I didn't see any reason to keep that.

### CODE REVIEW

**Comment by willdurand** [comment]
> `App` was creating a property called `this.header`, which was being set to a `ref` for the `Header` component. But I didn't see that property used anywhere.

Yep, this was used in the test suite a long time ago.

**Comment by willdurand** [comment]
> `App` also had a prop called `HeaderComponent` which seemed to allow a caller to pass a component other than `Header` into `App`. I don't see that used anywhere, not even in any tests, so I didn't see any reason to keep that.

I don't recall exactly why but it is probably *something something test-related*. It was similar for the `Footer` IIRC. I believe it is fine to remove all of that.

**Comment by bobsilverberg** [comment]
> 
> I don't recall exactly why but it is probably _something something test-related_. It was similar for the `Footer` IIRC. I believe it is fine to remove all of that.

Thanks @willdurand. I assumed it was for testing purposes, but none of the tests use it. I guess maybe a test was removed.

File: src/amo/components/ErrorPage/NotAuthorized/index.js [hunk 1]
```diff
@@ -32,21 +33,23 @@ export class NotAuthorizedBase extends React.Component<Props> {
...
+        <Page>
+          <Card
+            className="ErrorPage NotAuthorized"
+            header={i18n.gettext('Not Authorized')}
+          >
+            <p>
+              {i18n.gettext(`
+        Sorry, but you aren't authorized to access this page. If you
+        aren't signed in, try signing in using the link at the top
+        of the page.`)}
```
Comment (willdurand) [inline src/amo/components/ErrorPage/NotAuthorized/index.js]: Here and everywhere else, text blocks are misaligned (see now vs before).

File: src/amo/components/Page/index.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+export const PageBase = ({
+  _config = config,
+  children,
+  isHomePage = false,
+  location,
+}: InternalProps) => {
+  return (
+    <>
+      <Header isHomePage={isHomePage} location={location} />
+      <div className={makeClassName(isHomePage ? 'Page-homepage' : 'Page')}>
```
Comment (willdurand) [inline src/amo/components/Page/index.js]: I would always add `Page` as class name and only add `Page-is-homepage` when needed. That way we can have a default style, a class to rely on no matter what happens, and another style that overrides the default one when it is the homepage.

File: src/amo/pages/AddonVersions/index.js [hunk 3]
```diff
@@ -149,60 +150,63 @@ export class AddonVersionsBase extends React.Component<InternalProps> {
...
-                    }
-                    key={version.id}
-                    version={version}
-                  />
-                );
-              })}
-            </ul>
-          </CardList>
+      <Page>
+        {' '}
```
Comment (willdurand) [inline src/amo/pages/AddonVersions/index.js]: Why is this needed?

**Comment by willdurand** [comment]
Sorry, wrong button... I did not want to approve this patch right way. Instead, I left a suggestion and some comments.

**Comment by bobsilverberg** [comment]
> Except very few pages (for which we can certainly update the layout), all page components return a `div` with a class name. We could maybe allow `Page` to take a `className` prop and return a `<div />` with that class name. That would reduce the diff too, I think.

Ah, so you're saying have `Page` return the top level `div` instead of having the child create that `div`? Yeah, that sounds like a good idea.

File: src/amo/pages/AddonVersions/index.js (hunk 3 above)
Comment (bobsilverberg) [reply src/amo/pages/AddonVersions/index.js]: I used a "feature" of VSCode to select a block of code and then wrap it in a tag (to add the `<Page>` tag to each page component), but I later realized that it was sometimes doing weird things to the code that I was wrapping. This must have been an artifact of that.

**Comment by kumar303** [comment]
Be sure to double check every `component={}` assignment in [`<Routes>`](https://github.com/mozilla/addons-frontend/blob/9eb3c03f5399bed701c55375352258fd15e00638/src/amo/components/Routes/index.js). I don't see a Page wrapper around [`core/pages/error-simulation`](https://github.com/mozilla/addons-frontend/tree/master/src/core/pages/error-simulation). These are also shared by Discopane so it might get tricky. If it becomes tricky, this could be done in a follow-up as it's only a development feature.

**Comment by bobsilverberg** [comment]
> Be sure to double check every `component={}` assignment in [`<Routes>`](https://github.com/mozilla/addons-frontend/blob/9eb3c03f5399bed701c55375352258fd15e00638/src/amo/components/Routes/index.js). I don't see a Page wrapper around [`core/pages/error-simulation`](https://github.com/mozilla/addons-frontend/tree/master/src/core/pages/error-simulation). These are also shared by Discopane so it might get tricky. If it becomes tricky, this could be done in a follow-up as it's only a development feature.

I looked through https://github.com/mozilla/addons-frontend/blob/9eb3c03f5399bed701c55375352258fd15e00638/src/amo/components/Routes/index.js and I believe I have covered everything *except* the error simulators. If we can defer that to a follow-up that would be ideal.

**Comment by bobsilverberg** [comment]
Ok, this is ready for another look. Thanks for the suggestion @willdurand about replacing the `div` via `Page` - it has made the diff much cleaner!

**Comment by bobsilverberg** [comment]
@willdurand It looks like you're assigned to this now (it looks like you can blame @kumar303 for that 😉). I'd really like to get this into tomorrow's tag, so if you have some time today or tomorrow morning to look at it again that would be great. Thanks!

**Comment by bobsilverberg** [comment]
> > Be sure to double check every `component={}` assignment in [`<Routes>`](https://github.com/mozilla/addons-frontend/blob/9eb3c03f5399bed701c55375352258fd15e00638/src/amo/components/Routes/index.js). I don't see a Page wrapper around [`core/pages/error-simulation`](https://github.com/mozilla/addons-frontend/tree/master/src/core/pages/error-simulation). These are also shared by Discopane so it might get tricky. If it becomes tricky, this could be done in a follow-up as it's only a development feature.
> 
> I looked through [/src/amo/components/Routes/index.js@`9eb3c03`](https://github.com/mozilla/addons-frontend/blob/9eb3c03f5399bed701c55375352258fd15e00638/src/amo/components/Routes/index.js) and I believe I have covered everything _except_ the error simulators. If we can defer that to a follow-up that would be ideal.

I opened https://github.com/mozilla/addons-frontend/issues/8728 as a follow-up.

**Comment by willdurand** [comment]
> This patch could benefit from several pairs of eyes. I browsed through the site and I noticed that a few pages have incorrect styles. I included the dev versions (how they should look after each one).

I suggest we label the issue as "qa: extra" too.

**Comment by bobsilverberg** [comment]
> This patch could benefit from several pairs of eyes. I browsed through the site and I noticed that a few pages have incorrect styles. I included the dev versions (how they should look after each one).
> Screenshots

Thanks @kumar303. I did have to do something different for those pages, and I really should have compared the results to dev first. Sorry about that. I was feeling a bit of pressure knowing that we don't want what's currently in dev to go to prod this week. I'll take a look at these and see if I can figure out what's going wrong.

**Comment by willdurand** [comment]
> Thanks @kumar303. I did have to do something different for those pages, and I really should have compared the results to dev first. Sorry about that. I was feeling a bit of pressure knowing that we don't want what's currently in dev to go to prod this week. I'll take a look at these and see if I can figure out what's going wrong.

@bobsilverberg no rush, we'll have plenty of time to land this before tomorrow's tag. I planned to spend some time on it and to involve QA as early as possible tomorrow :)

**Comment by bobsilverberg** [comment]
Thanks @kumar303 and @willdurand. My apologies again for not checking everything myself before submitting that patch. I have fixed the issue @kumar303 reported and will push a new commit soon.

File: src/amo/components/Page/index.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+import AppBanner from 'amo/components/AppBanner';
+import Header from 'amo/components/Header';
+import Card from 'ui/components/Card';
+import type { ReactRouterLocationType } from 'core/types/router';
+
+import './styles.scss';
+
+type Props = {|
+  children: React.Node,
+  className?: string,
```
Comment (willdurand) [inline src/amo/components/Page/index.js]: I believe this should be named `componentClassName` because we usually use `className` for the top-level component that is rendered.

That makes me think that `componentProps` is not very explicit, although I get why it is called like this. Maybe we could find a better name. I suggest: `content`, so we could have `contentClassName`, `contentProps`, and maybe `useCardForContent`.

File: src/amo/components/Page/index.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+      <Header isHomePage={isHomePage} location={location} />
+      <div
+        className={makeClassName('Page', { 'Page-not-homepage': !isHomePage })}
+      >
+        {// Exclude the AppBanner from the home page if it will be
+        // included via HeroRecommendation.
+        (!isHomePage || !_config.get('enableFeatureHeroRecommendation')) && (
+          <AppBanner />
+        )}
+        {useCardComponent ? (
```
[the remark this answers is not in the document]
Comment (willdurand) [reply src/amo/components/Page/index.js]: If you follow my suggestion to use `content`, `Component` could be renamed to `Content`:

```jsx
<Content className={contentClassName} {...contentProps}>{children}</Content>
```

File: tests/unit/amo/components/TestPage.js [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+import Card from 'ui/components/Card';
+import {
+  createContextWithFakeRouter,
+  getFakeConfig,
+  shallowUntilTarget,
+} from 'tests/unit/helpers';
+
+describe(__filename, () => {
+  const render = (props = {}) => {
+    const allProps = { children: '', ...props };
```
Comment (willdurand) [inline tests/unit/amo/components/TestPage.js]: I'd suggest using a slightly more realistic `children` value here.

File: src/amo/components/App/index.js [hunk 7]
```diff
@@ -274,7 +232,6 @@ export class AppBase extends React.Component<Props> {
 export const mapStateToProps = (state: AppState) => ({
   authToken: state.api && state.api.token,
   clientApp: state.api.clientApp,
-  isHomePage: state.viewContext.context === VIEW_CONTEXT_HOME,
   lang: state.api.lang,
   userAgent: state.api.userAgent,
 });
```
[the remark this answers is not in the document]
Comment (bobsilverberg) [reply src/amo/components/App/index.js]: I'm still not sure *why* we need it. I found this: https://github.com/ReactTraining/react-router/blob/e634f0bad7796d128a4c4e2c4678487dd7256be2/packages/react-router/docs/api/withRouter.md#important-note, but I don't think that really explains it. I can say that if we don't include `withRouter` in `App`, then the app no longer responds to location changes. The URL changes, but nothing happens in the app.

I tried for a long time to figure out how to write a test to check that `App` has been wrapped in `withRouter`. I came up with one solution, but it's kind of odd.

File: src/amo/components/ErrorPage/NotAuthorized/index.js [hunk 8]
```diff
@@ -32,9 +32,10 @@ export class NotAuthorizedBase extends React.Component<Props> {
     /* eslint-disable react/no-danger */
     return (
       <NestedStatus code={401}>
-        <Card
+        <Page
           className="ErrorPage NotAuthorized"
-          header={i18n.gettext('Not Authorized')}
+          componentProps={{ header: i18n.gettext('Not Authorized') }}
+          useCardComponent
```
[the remark this answers is not in the document]
Comment (bobsilverberg) [reply src/amo/components/ErrorPage/NotAuthorized/index.js]: I liked it better too, but it was broken. I now realize I can bring it back by just changing the syntax I was using, so I've done that. Thanks!

File: src/amo/components/Page/index.js [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+  componentProps = {},
+  isHomePage = false,
+  location,
+  useCardComponent = false,
+}: InternalProps) => {
+  return (
+    <>
+      <Header isHomePage={isHomePage} location={location} />
+      <div
+        className={makeClassName('Page', { 'Page-not-homepage': !isHomePage })}
```
[the remark this answers is not in the document]
Comment (bobsilverberg) [reply src/amo/components/Page/index.js]: I agree with your statement, but the issue is that the only difference between the homepage and non-homepage style is that the non-homepage defines one extra property. So it seemed easiest to just add a style for that, and then the extra property could be applied. I suppose I could have made it a negative rule in the stylesheet, but that seems more difficult to reason about than what I've done here. WDYT?

File: src/amo/components/Page/index.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (bobsilverberg) [reply src/amo/components/Page/index.js]: I have addressed this by going back to allowing the caller to specify a component type to use.

File: src/amo/components/Page/index.js [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+      >
+        {// Exclude the AppBanner from the home page if it will be
+        // included via HeroRecommendation.
```
[the remark this answers is not in the document]
Comment (bobsilverberg) [reply src/amo/components/Page/index.js]: Should a comment be added to `AppBanner` or to `HeroRecommendation`? If I am understanding the reason for this comment, it seems like it would be more beneficial in `HeroRecommendation`, correct?

File: src/amo/components/Page/index.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (willdurand) [reply src/amo/components/Page/index.js]: Oh I see. Okay, let's keep it as is

File: src/amo/components/Page/index.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (willdurand) [reply src/amo/components/Page/index.js]: Ah. Correct.

**Comment by willdurand** [comment]
> The reason `withRouter` is needed here is because the component is wrapped in `connect()` and that _blocks_ re-renders in all cases except when `mapStateToProps()` returns differing component values. This is how `connect()` works everywhere, on every component. Thus, I don't think it's so helpful to add a comment or a test because that's just how it works and we need to be aware of it.

Are you referring to this? https://reacttraining.com/react-router/web/guides/redux-integration/blocked-updates

If so I suggest we leave a comment with this link because that's a special case in our app.

File: src/amo/components/Page/index.js [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+|};
+
+export const PageBase = ({
+  _config = config,
+  children,
+  isHomePage = false,
+  location,
+}: InternalProps) => {
+  return (
+    <div className="App-amo">
```
Comment (willdurand) [inline src/amo/components/Page/index.js]: I guess we could rename the class names now.

File: src/amo/components/Page/index.js [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+      <InfoDialog />
+
+      <Header isHomePage={isHomePage} location={location} />
+
+      <div className="App-content">
```
Comment (willdurand) [inline src/amo/components/Page/index.js]: Has the style moved to `Page/styles.scss` too?

File: tests/unit/amo/components/TestApp.js [hunk 13]
```diff
@@ -119,13 +114,6 @@ describe(__filename, () => {
     });
   });
 
-  it('renders an ErrorPage component', () => {
```
Comment (kumar303) [inline tests/unit/amo/components/TestApp.js]: You can put this test back now

File: src/amo/components/Page/index.js (hunk 12 above)
Comment (bobsilverberg) [reply src/amo/components/Page/index.js]: I'll move them

File: tests/unit/amo/pages/TestLandingPage.js [hunk 14]
```diff
@@ -253,7 +254,9 @@ describe(__filename, () => {
     const root = render({
       params: { visibleAddonType: getVisibleAddonType(ADDON_TYPE_EXTENSION) },
     });
-    expect(root).not.toHaveClassName('.LandingPage--theme');
+    expect(root.find('.LandingPage')).not.toHaveClassName(
```
Comment (kumar303) [inline tests/unit/amo/pages/TestLandingPage.js]: Good catch on updating the `not` selectors 😬

File: src/amo/components/ErrorPage/NotAuthorized/index.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (kumar303) [reply src/amo/components/ErrorPage/NotAuthorized/index.js]: Whoops, ignore this. I guess it was sitting in a draft state from an earlier review request.

File: tests/unit/amo/components/TestApp.js (hunk 13 above)
Comment (bobsilverberg) [reply tests/unit/amo/components/TestApp.js]: The test was checking that we could pass in a custom `ErrorPage` component, but that was never used anywhere other than this test, so I removed it, much like I removed `HeaderComponent` and `FooterComponent`. Without the ability to override `ErrorPage`, this test seems pointless. WDYT?

Comment (kumar303) [reply tests/unit/amo/components/TestApp.js]: Oh, whoops, I see. My best guess is that it was added to prove dependency injection before we could do shallow rendering. Sure, you can remove it. I see that dependency injection is longer possible, which is fine: https://github.com/mozilla/addons-frontend/pull/8717/files#diff-53db53641895141db51830eafab741bcL37

Comment (bobsilverberg) [reply tests/unit/amo/components/TestApp.js]: Correct. I removed that "feature".

**Comment by bobsilverberg** [comment]
> I did a final test of the site and everything is working.
> 
> I wanted to check that 500/400 errors were still displaying correctly. They are but I had to fix the simulate pages and I did so with this small patch. I think it would be fine to ship this if you want to include it. They are just simulator pages so it's not a big deal.

Ah, I didn't think of doing that! Neat trick. I will make that change and then ship it. Btw, you can also go to http://localhost:3000/en-US/firefox/401/, but maybe that is testing something different?

**Comment by kumar303** [comment]
The simulators will exercise the error boundary which -- in real life -- could get take over if some nested component throws an error. It's a different code path than routing directly to the error templates.
